envoy/matcher/matcher.h

```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
 // Conceptually the data availability should start at being NotAvailable, transition to
 // MoreDataMightBeAvailable (optional, this doesn't make sense for all data) and finally
 // AllDataAvailable as the data becomes available.
-struct DataInputGetResult {
-  enum class DataAvailability {
-    // The data is not yet available.
-    NotAvailable,
-    // Some data is available, but more might arrive.
-    MoreDataMightBeAvailable,
-    // All the data is available.
-    AllDataAvailable
-  };
-
+template <class ResultDataType = std::string> struct DataInputGetResult {
   DataAvailability data_availability_;
   // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
   // per data_availability_) or because no value was extracted. For example, consider a DataInput
   // which attempts to look a key up in the map: if we don't have access to the map yet, we return
   // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
   // map, we return absl::nullopt with AllDataAvailable.
-  absl::optional<std::string> data_;
+  absl::optional<ResultDataType> data_;
 
   // For pretty printing.
   friend std::ostream& operator<<(std::ostream& out, const DataInputGetResult& result) {
-    out << "data input: " << (result.data_ ? result.data_.value() : "n/a");
+    if (result.data_.has_value()) {
+      out << "data input: " << result.data_.value();
+    } else {
+      out << "data input: n/a";
+    }
+
     switch (result.data_availability_) {
-    case DataInputGetResult::DataAvailability::NotAvailable:
+    case DataAvailability::NotAvailable:
       out << " (not available)";
       break;
-    case DataInputGetResult::DataAvailability::MoreDataMightBeAvailable:
+    case DataAvailability::MoreDataMightBeAvailable:
       out << " (more data available)";
       break;
-    case DataInputGetResult::DataAvailability::AllDataAvailable:;
+    case DataAvailability::AllDataAvailable:;
     }
     return out;
   }
 };
 
+// Template deduction guide for DataInputGetResult.
+template <class ResultDataType>
```

Comment:
The other alternative here is we make the input type some oneof union. It could be a string as today, or it could be a protobuf dynamic type, which supports a fairly wide variety of data models. The main downside there is performance, which we probably should be sensitive to on the match path, so I think I'm convinced that doing this at C++ level makes sense. Not certain though this needs to be templates, will defer to @snowp.

Reply:
In my last attempt, I explored `std::any` and `std::variant`. The first one is a terrible choice, since it's hard to work with. Variant could work but it makes the input structure wide and not open-ended. Ultimately, the main difference is in the way matchers adapt to inputs of certain types. It can be done with a big switch on `input.type` or via template substitution `template <Type> match(DataInputGetResult<Type>)`. If we go with templates, I want to make sure we have a clear way to support integers - I think template specialization for each numeric type is too error-prone.

Reply:
Thank you both for the review and inputs!

Re htuch: at C++ level,  I think template is also the one aligned with existing matching API, similar to how `DataType` template is used.

Re kyessenov: I mostly focused on supporting CEL though I did test with INT in my prototype.

I am definitely open to other ideas. It is probably better to start with the design doc which includes whole picture and more details(e.g., use example to both data input and input matcher) of this approach. I will share with you later for discussion and review.

Reply:
Originally we went with templates because it worked nicely with the way factory registrations work: `RequestHeaderInput` would sub class `DataInput<HttpData>` and be a distinct registration that wouldn't work with something looking for `DataInput<TcpData>` registrations. One of the complications from this though was that we ended up with a lot of repeated registrations whenever we wanted to register the factories for different kind of inputs. 

So learning from that, it would be nice that we try to avoid making extension registration/implementation even more complicated as we're starting to support multiple output types. 

It might be that we find that templates is the only efficient way of doing this (std::variant has some overhead), but I'd be interested to see what it would look like with std::variant. I think you could get pretty far with some helper functions/base classes that would hide the std::variant unpacking from the extension writer

tl;dr a doc would be appreciated :)

Reply:
From the point of view of consumers of inputs, they are mostly interested in a projection to the desired type, e.g. to string or to int or to CEL environment. So we can use regular interfaces to express this without templates I think. I've looked into how CEL represents values in C++ - there's a lot of useful design ideas there.

Another important property we probably want - either make Input non-copyable or make it trivially copyable by moving the storage elsewhere. A large object like CEL environment is undesirable to copy.

Reply:
I think runtime check (like a big switch on input.type mentioned above) is not good for performance.  

And yes, I did change the interface of input matcher to take in `const ref` instead of `absl::optional` exactly to avoid copy of large object like cel environment.(will be shown in next PR though as this PR only has change for dataInput) 

I have shared the design doc with you all. Please take a look. Thank you!

Reply:
Putting aside performance for a second, I'm wondering why we can't just use a solution similar to `FilterState` and at config validation time some RTTI to validate compatibility of input-matcher types? Then maybe variant is interesting as some kind of optimization for unboxed types but not inherently needed?

Looking at the code, I'm definitely in agreement with @snowp that we don't want to increase the level of template type complexity if we can avoid it, it's already quite hard to follow the types and instance-template relationships when paging back in after some time.